### PR TITLE
feat: Implement backup and restore commands

### PR DESCRIPTION
**Pull request by mmorhun**
Signed-off-by: Mykola Morhun <mmorhun@redhat.com>

### What does this PR do?
Add `server:backup` and `server:restore` commands

### What issues does this PR fix or reference?
https://github.com/eclipse/che/issues/19932

#### Internal backup server
1. Deploy Eclipse Che and make some changes
2. Request backup by executing `chectl server:backup` and wait until it finishes
3. Change / harm installation
4. Request restore by executing `chectl server:restore` and wait until it finishes
5. Check original config is applied

#### An external backup server
1. Deploy Eclipse Che and make some changes
2. Request backup by executing `chectl server:backup -r rest:my.server.net:8000/che-backup` and wait until it finishes
3. Change / harm installation
4. Request restore by executing `chectl server:restore` and wait until it finishes
5. Check original config is applied

### PR Checklist

[As the author of this Pull Request I made sure that:](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#pull-request-template-and-its-checklist)

- [x] [The Eclipse Contributor Agreement is valid](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-eclipse-contributor-agreement-is-valid)
- [x] [Code produced is complete](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-produced-is-complete)
- [x] [Code builds without errors](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#code-builds-without-errors)
- [x] [Tests are covering the bugfix](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#tests-are-covering-the-bugfix)
- [ ] [The repository devfile is up to date and works](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#the-repository-devfile-is-up-to-date-and-works)
- [x] [Sections `What issues does this PR fix or reference` and `How to test this PR` completed](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#sections-what-issues-does-this-pr-fix-or-reference-and-how-to-test-this-pr-completed)
- [ ] [Relevant user documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [Relevant contributing documentation updated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#relevant-contributing-documentation-updated)
- [ ] [CI/CD changes implemented, documented and communicated](https://github.com/eclipse/che/blob/master/CONTRIBUTING.md#cicd-changes-implemented-documented-and-communicated)

### Reviewers

Reviewers, please comment how you tested the PR when approving it.

### CODE REVIEW

File: src/commands/server/backup.ts [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+    'chectl server:backup',
+    '# Create and use configuration for REST backup server:\n' +
+    'chectl server:backup -r rest:http://my-sert-server.net:4000/che-backup -p repopassword',
+    '# Create and use configuration for AWS S3 (and API compatible) backup server (bucket should be precreated):\n' +
+    'chectl server:backup -r s3:s3.amazonaws.com/bucketche -p repopassword',
+    '# Create and use configuration for SFTP backup server:\n' +
+    'chectl server:backup -r sftp:user@my-server.net:/srv/sftp/che-data -p repopassword',
+  ]
+
+  static flags: flags.Input<any> = {
```
Comment (tolusha) [inline src/commands/server/backup.ts:109]: I don't like an idea to provide password via cli. Can we found a better way?

File: src/commands/server/restore.ts [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+    'snapshot-id': string({
+      char: 's',
+      description: 'ID of a snapshot to restore from',
+      required: false,
+    }),
+  }
+
+  async run() {
+    const { flags } = this.parse(Restore)
+    flags.chenamespace = flags.chenamespace || DEFAULT_CHE_NAMESPACE
```
Comment (tolusha) [inline src/commands/server/restore.ts:67]: The same remarks as above.

File: src/commands/server/restore.ts [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+    const ctx = await ChectlContext.initAndGet(flags, this)
+
+    await this.config.runHook(DEFAULT_ANALYTIC_HOOK_NAME, { command: Restore.id, flags })
+
+    const tasks = new Listr([], ctx.listrOptions)
+    const apiTasks = new ApiTasks()
+    tasks.add(apiTasks.testApiTasks(flags))
+    tasks.add(this.getRestoreTasks(flags))
+    // Export self-signed CA cert if any
+    flags.tls = true
```
Comment (tolusha) [inline src/commands/server/restore.ts:77]: Not sure if I understand this.

File: src/api/backup-restore.ts [hunk 4]
```diff
@@ -0,0 +1,281 @@
...
+}
+export interface AwsS3BackupServerCredentials {
+  awsAccessKeyId: string
+  awsSecretAccessKey: string
+}
+export interface SftpBackupServerCredentials {
+  sshKey: string
+}
+
+export const BACKUP_CR_NAME = 'chectl-che-backup'
```
[the remark this answers is not in the document]
Comment (mmorhun) [reply src/api/backup-restore.ts:55]: To be sure we do not override user's CR.

File: src/api/backup-restore.ts [hunk 5]
```diff
@@ -0,0 +1,281 @@
...
+export const REST_SERVER_CREDENTIALS_SECRET_NAME = 'chectl-backup-rest-server-credentials'
+export const AWS_CREDENTIALS_SECRET_NAME = 'chectl-aws-credentials'
+export const SSH_KEY_SECRET_NAME = 'chectl-backup-sftp-server-key'
+
+/**
+ * Detects backup server type. Returns empty if there is no type specified or type is invalid.
+ * @param url full url to backup server including restic protocol, e.g. sftp://url
+ */
+export function getBackupServerType(url: string): BackupServerType {
+  if (url.startsWith('rest:') || url.startsWith('rest://')) {
```
[the remark this answers is not in the document]
Comment (mmorhun) [reply src/api/backup-restore.ts:70]: No :)

File: src/commands/server/restore.ts (hunk 3 above)
Comment (mmorhun) [reply src/commands/server/restore.ts:77]: `retrieveCheCaCertificateTask` has `enabled: () => flags.tls && flags.installer !== 'helm'`
That's why I added `tls` flag.

File: src/api/backup-restore.ts [hunk 6]
```diff
@@ -0,0 +1,281 @@
...
+}
+
+/**
+ * Submits backup of Che installation task.
+ * @param namespace namespace in which Che is installed
+ * @param backupServerConfig backup server configuration data or name of the config CR
+ */
+export async function requestBackup(namespace: string, backupServerConfig?: BackupServerConfig | string): Promise<V1CheClusterBackup> {
+  const kube = new KubeHelper()
+  const backupServerConfigName = await getBackupServerConfigurationName(namespace, backupServerConfig)
```
[the remark this answers is not in the document]
Comment (mmorhun) [reply src/api/backup-restore.ts:87]: No, it is optional for `getBackupServerConfigurationName`

File: src/commands/server/backup.ts [hunk 7]
```diff
@@ -0,0 +1,292 @@
...
+    [SSH_KEY_FILE_KEY]: sshKeyFile,
+    [AWS_ACCESS_KEY_ID_KEY]: awsAccessKeyId,
+    [AWS_SECRET_ACCESS_KEY_KEY]: awsSecretAccessKey,
+    [BACKUP_SERVER_CONFIG_CR_NAME_KEY]: backupServerConfigName,
+  }
+
+  async run() {
+    const { flags } = this.parse(Backup)
+    flags.chenamespace = flags.chenamespace || DEFAULT_CHE_NAMESPACE
+    const ctx = await ChectlContext.initAndGet(flags, this)
```
[the remark this answers is not in the document]
Comment (mmorhun) [reply src/commands/server/backup.ts:126]: I moved it, but why? I see no obvious reason.

File: src/commands/server/backup.ts (hunk 1 above)
Comment (mmorhun) [reply src/commands/server/backup.ts:109]: This is an example.
In general, it is possible to do:
```sh
chectl server:backup -r sftp:user@my-server.net:/srv/sftp/che-data -p $REPO_PASSWORD
```
or
```sh
export BACKUP_REPOSITORY_PASSWORD=password
chectl server:backup -r sftp:user@my-server.net:/srv/sftp/che-data
```

Do you want me to change the example to ` -p $REPO_PASSWORD` ?

File: src/api/backup-restore.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (mmorhun) [reply src/api/backup-restore.ts:55]: Also to see that the CR is created by chectl

File: src/api/backup-restore.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (tolusha) [reply src/api/backup-restore.ts:70]: my mistake, vice versa then ;)

File: src/api/backup-restore.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (tolusha) [reply src/api/backup-restore.ts:55]: It is also user's CR but created with chectl.
if we need to mark CR with something then add this to annotation. 
so, pls remove that prefix.

File: src/commands/server/backup.ts (hunk 7 above)
[the remark this answers is not in the document]
Comment (tolusha) [reply src/commands/server/backup.ts:125]: Agree there is no actual necessity but let's follow this pattern like in other commands

File: src/commands/server/restore.ts (hunk 3 above)
Comment (tolusha) [reply src/commands/server/restore.ts:78]: pls remove retrieving certificate.
there is a dedicated command cacert:export

Comment (mmorhun) [reply src/commands/server/restore.ts:78]: ok

File: src/api/backup-restore.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (mmorhun) [reply src/api/backup-restore.ts:70]: But I need the expression to be true for, say, `rest:server.net`.

[the remark this answers is not in the document]
Comment (mmorhun) [reply src/api/backup-restore.ts:70]: Thanks, will fix

File: src/api/kube.ts [hunk 8]
```diff
@@ -1710,6 +1711,69 @@ export class KubeHelper {
     }
   }
 
+  async replaceCheGroupCr(crYaml: any, kindPlural: string): Promise<any> {
```
[the remark this answers is not in the document]
Comment (mmorhun) [reply src/api/kube.ts:1714]: It doesn't work. Replaces doesn't recreate the object, so reconcile loop isn't triggered, so no new backups created....

[the remark this answers is not in the document]
Comment (mmorhun) [reply src/api/kube.ts:1714]: Reverting

**Comment by mmorhun** [comment]
Rebased

**Review by tolusha — APPROVED** [review]

**Pull request merged**